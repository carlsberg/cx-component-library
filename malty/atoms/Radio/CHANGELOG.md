# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.1.0 (2024-05-31)

### Bug Fixes

- **DSM-364:** changed id prop ([#312](https://github.com/CarlsbergGBS/cx-component-library/issues/312)) ([20ed3b1](https://github.com/CarlsbergGBS/cx-component-library/commit/20ed3b12385242d90fbedc308818302d92ac5226))
- **radio:** [CADI-16915] revert change of Radio selected prop ([#946](https://github.com/CarlsbergGBS/cx-component-library/issues/946)) ([908ae7b](https://github.com/CarlsbergGBS/cx-component-library/commit/908ae7ba6883d2b71c9fd22c8f4b597c7f511a4e))
- **radio:** [DSM-938] remove hover behaviour when the component is disabled ([#825](https://github.com/CarlsbergGBS/cx-component-library/issues/825)) ([c6b8bd0](https://github.com/CarlsbergGBS/cx-component-library/commit/c6b8bd0042931bf52b237c9b61954d8c8eceb0f8))
- Remove typography provider from every component ([#383](https://github.com/CarlsbergGBS/cx-component-library/issues/383)) ([5d05680](https://github.com/CarlsbergGBS/cx-component-library/commit/5d05680e40d3da5bb3042bd8191427a70606afc3))
- reverted changes ([#396](https://github.com/CarlsbergGBS/cx-component-library/issues/396)) ([ecfcf57](https://github.com/CarlsbergGBS/cx-component-library/commit/ecfcf573863253fea16e24764b6f452c3cc4f215))
- Run auto fixable lint based on new `eslint` configuration ([#501](https://github.com/CarlsbergGBS/cx-component-library/issues/501)) ([9ee94cc](https://github.com/CarlsbergGBS/cx-component-library/commit/9ee94cc2cd407f717e62d8857b5f20a74e7bbec4))

### Features

- add tsconfig files and update package.json files for each component ([3bfc644](https://github.com/CarlsbergGBS/cx-component-library/commit/3bfc644e1cfc9dbb7cf7a0469e25fce055b53240))
- additional stories ([#291](https://github.com/CarlsbergGBS/cx-component-library/issues/291)) ([438af3b](https://github.com/CarlsbergGBS/cx-component-library/commit/438af3b969e0f5b4b908436ee58b93c4b9ddc97c))
- change organization name ([a0aa428](https://github.com/CarlsbergGBS/cx-component-library/commit/a0aa428f43138af5707a3ff4c0e36b7c056f02e6))
- **DSM-228:** Added rating component ([#379](https://github.com/CarlsbergGBS/cx-component-library/issues/379)) ([ab47354](https://github.com/CarlsbergGBS/cx-component-library/commit/ab47354db803fb051a1f87d66ee34966698df8bd))
- **DSM-248:** Headline HTML tags ([#184](https://github.com/CarlsbergGBS/cx-component-library/issues/184)) ([305f5e0](https://github.com/CarlsbergGBS/cx-component-library/commit/305f5e0fa92d5813f4877b338ad827a8c347d452))
- **DSM-362:** snapshots removed & unit test updates ([#307](https://github.com/CarlsbergGBS/cx-component-library/issues/307)) ([78423eb](https://github.com/CarlsbergGBS/cx-component-library/commit/78423eb5cfe04fe8a3a7af1f5af074ac6d35e235))
- **DSM-368:** Required prop & label component ([#384](https://github.com/CarlsbergGBS/cx-component-library/issues/384)) ([983bd78](https://github.com/CarlsbergGBS/cx-component-library/commit/983bd78c8f526a8faa2df601bf918b40219820f2))
- **DSM-439:** Change font-family behavior ([#470](https://github.com/CarlsbergGBS/cx-component-library/issues/470)) ([ab6dad4](https://github.com/CarlsbergGBS/cx-component-library/commit/ab6dad414c42c4343de3de6abaa4cd3ff98cfca1))
- **global:** [DSM-784] add space props in all components ([#842](https://github.com/CarlsbergGBS/cx-component-library/issues/842)) ([9f03f59](https://github.com/CarlsbergGBS/cx-component-library/commit/9f03f59e7518a99abcea06bacbee989d1ddb0465))
- **global:** migrate from bit to lerna ([a0110c8](https://github.com/CarlsbergGBS/cx-component-library/commit/a0110c8831370dc762c193b17cc593eed381f990))
- **global:** Remove `react/jsx-props-no-spreading` comments ([#558](https://github.com/CarlsbergGBS/cx-component-library/issues/558)) ([823444c](https://github.com/CarlsbergGBS/cx-component-library/commit/823444ce0800bc7f49b584de59d247eb8497f559))
- **global:** update storybook files to be compliant with v7 ([#970](https://github.com/CarlsbergGBS/cx-component-library/issues/970)) ([c61d4b4](https://github.com/CarlsbergGBS/cx-component-library/commit/c61d4b40a0755becf942ad3f28758a159f8c54e4))
- **radio:** [DSM-725] add tests ([#720](https://github.com/CarlsbergGBS/cx-component-library/issues/720)) ([1a670db](https://github.com/CarlsbergGBS/cx-component-library/commit/1a670db68cee4793325725bca3917e0abfe2d27d))
- **radio:** [DSM-942] add readonly ([#848](https://github.com/CarlsbergGBS/cx-component-library/issues/848)) ([a38744d](https://github.com/CarlsbergGBS/cx-component-library/commit/a38744dbda2b7f48f65bb5ae17e5ad31168ecc86))
- remove empty composition files ([f8173f4](https://github.com/CarlsbergGBS/cx-component-library/commit/f8173f4a2ecbf80bb7b6ffe848c023ae31819c2d))
- Use new `eslint-config-carlsberg` with streamlined way to create components ([#559](https://github.com/CarlsbergGBS/cx-component-library/issues/559)) ([a459332](https://github.com/CarlsbergGBS/cx-component-library/commit/a45933215bc0b523220743f17d988d7f1ad5fce6))
