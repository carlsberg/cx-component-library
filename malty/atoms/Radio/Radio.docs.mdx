---
labels: ['react', 'typescript', 'text']
description: 'A Radio component.'
---
