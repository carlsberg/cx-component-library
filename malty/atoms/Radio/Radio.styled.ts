import styled, { css } from 'styled-components';

export const StyledRadio = styled.input`
  display: inline-block;
  height: ${({ theme }) => theme.variables.radio.size.value}px;
  width: ${({ theme }) => theme.variables.radio.size.value}px;
  margin: 0;

  position: relative;

  cursor: pointer;

  &:before {
    content: '';
    transition: transform 0.4s cubic-bezier(0.45, 1.8, 0.5, 0.75);
    position: absolute;
    top: 50%;
    transform: translateY(-50%) scale(0, 0);
    left: 5px;
    z-index: 1;

    width: ${({ theme }) => theme.variables.radio.inputSelected.value}px;
    height: ${({ theme }) => theme.variables.radio.inputSelected.value}px;

    background: ${({ theme }) => theme.color.black.value};

    border-radius: ${({ theme }) => theme.variables.radio.borderRadius.value}%;
  }

  &:checked {
    &:before {
      transform: translateY(-50%) scale(1, 1);
    }
  }

  &:after {
    content: '';

    position: absolute;

    width: ${({ theme }) => theme.variables.radio.size.value}px;
    height: ${({ theme }) => theme.variables.radio.size.value}px;
    box-sizing: border-box;
    background: ${({ theme }) => theme.color.white.value};

    border: 2px solid ${({ theme }) => theme.color.default.value};
    border-radius: ${({ theme }) => theme.variables.radio.borderRadius.value}%;
  }

  ${({ disabled }) =>
    disabled &&
    css`
      pointer-events: none;
      &:before {
        background: ${({ theme }) => theme.color.button.primaryDisable.value};
      }
      &:after {
        background: ${({ theme }) => theme.color.white.value};
        border: 2px solid ${({ theme }) => theme.color.button.primaryDisable.value};
      }
    `}
`;

export const StyledLabel = styled.label<{
  disabled?: boolean;
}>`
  color: ${({ theme }) => theme.color.default.value};
  font-size: ${({ theme }) => theme.typography.text['medium-small']['font-size'].value}px;
  line-height: ${({ theme }) => theme.typography.text['medium-small']['line-height'].value}px;
  padding-left: ${({ theme }) => theme.variables.radio.paddingLeft.value}px;
  font-weight: 400;
  cursor: pointer;
  ${({ disabled }) =>
    disabled &&
    css`
      color: ${({ theme }) => theme.color.information.disable.value};
      pointer-events: none;
    `}
`;

export const StyledError = styled.label`
  font-family: inherit;
  color: ${({ theme }) => theme.color.system.failStrong.value};
  font-size: ${({ theme }) => theme.typography.information.tiny['font-size'].value}px;
  font-weight: bold;
  line-height: ${({ theme }) => theme.typography.information.tiny['line-height'].value}px;
  letter-spacing: 0;
`;
export const StyledRadioContainer = styled.div`
  display: flex;
  align-items: center;
`;
