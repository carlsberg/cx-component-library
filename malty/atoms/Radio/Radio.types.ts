export interface RadioProps extends React.HTMLAttributes<HTMLInputElement> {
  value: string | number;
  name: string;
  label: string;
  error?: string;
  selected?: boolean;
  disabled?: boolean;
  onValueChange?: (value: string | number) => void;
  required?: boolean;
  dataTestId?: string;
  readOnly?: boolean;
}
