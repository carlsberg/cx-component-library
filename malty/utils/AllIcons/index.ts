export { allIcons, allIconsStoryOptions } from './AllIconsMap';
