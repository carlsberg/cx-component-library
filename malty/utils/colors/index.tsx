const rgbToHex = (value: string): string =>
  value.replace(/rgb\((\d+),\s*(\d+),\s*(\d+)\)/gi, (_, r, g, b) => {
    const toHex = (n: string) => parseInt(n, 10).toString(16).padStart(2, '0');
    return `#${toHex(r)}${toHex(g)}${toHex(b)}`;
  });

export { rgbToHex };
