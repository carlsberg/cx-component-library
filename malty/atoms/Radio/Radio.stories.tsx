import { Story } from '@storybook/react';
import React from 'react';
import { RadioProps } from '.';
import { Radio as RadioComponent } from './Radio';

export default {
  title: 'Forms/Radio',
  component: RadioComponent,
  parameters: {
    importObject: 'Radio',
    importPath: '@carlsberggroup/malty.atoms.radio',
    variants: ['disabled'],
  },
  argTypes: {
    label: {
      description: 'label text radio',
      control: 'text',
    },
    error: {
      description: 'Error message to be displayed when error is present.',
      control: 'text',
    },
    value: {
      description: 'Value of Radio component',
      control: 'text',
    },

    selected: {
      description: 'If Radio component is selected or not',
      control: 'boolean',
    },
    disabled: {
      control: 'boolean',
      description: 'Input state, when disabled it is read-only.',
    },
    name: {
      description: 'Name attribute of the radio element.',
    },
    required: {
      control: 'boolean',
      description: 'Makes the radio input required to fill',
    },
    onValueChange: {
      description: 'Function to be executed when radio state changes',
    },
  },
};
const Template: Story<RadioProps> = ({
  value,
  label,
  selected,
  onValueChange,
  name,
  error,
  disabled,
  required,
}) => (
  <RadioComponent
    id={value.toString()}
    value={value}
    label={label}
    selected={selected}
    onValueChange={onValueChange}
    name={name}
    error={error}
    disabled={disabled}
    required={required}
  />
);
export const Radio = Template.bind({});

const params = new URLSearchParams(window.location.search);
const variant = params.get('variant');

switch (variant) {
  case 'disabled':
    Radio.args = {
      label: 'Option 1',
      name: 'radioOptions',
      value: 'Option 1',
      selected: true,
      disabled: true,
      required: false,
    };
    break;
  default:
    Radio.args = {
      label: 'Option 1',
      name: 'radioOptions',
      value: 'Option 1',
      selected: true,
      disabled: false,
      required: false,
    };
    break;
}
