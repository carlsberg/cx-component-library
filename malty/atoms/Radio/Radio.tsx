import { globalTheme as defaultTheme, TypographyProvider } from '@carlsberggroup/malty.theme.malty-theme-provider';
import React, { useContext } from 'react';
import { ThemeContext } from 'styled-components';
import { StyledError, StyledLabel, StyledRadio, StyledRadioContainer } from './Radio.styled';
import { RadioProps } from '.';

export const Radio = ({ value, label, onValueChange, selected, error, name, disabled, ...props }: RadioProps) => {
  const theme = useContext(ThemeContext) || defaultTheme;

  const handleValueChange = (e: { target: { value: string | number } }) => {
    onValueChange(e.target.value);
  };

  return (
    <TypographyProvider>
      <StyledRadioContainer theme={theme}>
        <StyledRadio
          id={value.toString()}
          checked={selected}
          value={value}
          theme={theme}
          type="radio"
          name={name}
          onChange={handleValueChange}
          disabled={disabled}
          // eslint-disable-next-line react/jsx-props-no-spreading
          {...props}
        />
        <StyledLabel htmlFor={value.toString()} disabled={disabled} theme={theme}>
          {label}
        </StyledLabel>
      </StyledRadioContainer>
      {error && <StyledError theme={theme}>{error}</StyledError>}
    </TypographyProvider>
  );
};
