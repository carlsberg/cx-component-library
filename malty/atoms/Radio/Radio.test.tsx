import { render } from '@carlsberggroup/malty.utils.test';
import { fireEvent, screen } from '@testing-library/react';
import React from 'react';
import { Radio } from './Radio';

describe('radio', () => {
  it('renders elements', () => {
    const mockFn = jest.fn();
    render(
      <Radio name="radio" label="Label text" error="Error text" value="Test value" onValueChange={mockFn} selected />
    );
    expect(screen.getByLabelText('Label text')).toBeInTheDocument();
    expect(screen.getByText('Error text')).toBeInTheDocument();
    expect(screen.getByDisplayValue('Test value')).toBeInTheDocument();
  });

  it('calls function on click', () => {
    const mockFn = jest.fn();
    render(<Radio label="Label text" error="Error text" value="Test value" onValueChange={mockFn} name="radio" />);
    const radio = screen.getByDisplayValue('Test value');
    fireEvent.click(radio);
    expect(mockFn).toHaveBeenCalledTimes(1);
  });
});
