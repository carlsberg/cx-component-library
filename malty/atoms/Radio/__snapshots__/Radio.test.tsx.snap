// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`radio matches snapshot 1`] = `
<span
  className="sc-bdvvtL iouZXA"
>
  <div
    className="sc-eCImPb cgYwop"
  >
    <input
      checked={true}
      className="sc-gsDKAQ kbuuGX"
      id="test"
      name="radio"
      onChange={[Function]}
      type="radio"
      value="test"
    />
    <label
      className="sc-dkPtRN ljICfl"
      htmlFor="test"
    >
      Label text
    </label>
  </div>
</span>
`;
